### DISCUSSION TITLE
Template fails to load if it contains "web" commands when internet is unavailable

### DISCUSSION THREAD

**Discussion by Lurux** [SilentVoid13/Templater#779] [Help]
- OS: Linux Mint 20.3 ≈ Ubuntu 20.04
- Templater version: 1.12.0
- Obsidian version: 0.13.33
- Templater settings: using `tp.web.random_picture()` or `tp.web.daily_quote()` as part of a template

<details><summary>Template code</summary>

```
[[Todo]] - [[Daily/<% tp.date.now("D-MM-Y") %> | Daily note]]


**Random notes**

<%*
let notes = app.vault.getMarkdownFiles();

for(let i = 0; i < 5; i++) {
	const index = Math.floor(Math.random() * notes.length);
	tR += `\n- [[${ notes[index].basename }]]`;
}
%>

**Random quote**

<% tp.web.daily_quote() %>

**Random image**

<% tp.web.random_picture() %>
```
</details>

**Describe the bug**

When trying to load that template while internet is disconnected or unavailable, the whole template fails to load with message "template parsing error, aborting"

<details><summary>Console errors</summary>

```
GET https://api.quotable.io/random net::ERR_INTERNET_DISCONNECTED
Templater Error: Template parsing error, aborting. Failed to fetch
GET https://source.unsplash.com/random/? net::ERR_INTERNET_DISCONNECTED
```
</details>

(Do note the template works as expected without internet if I remove the two web commands)

**Expected behavior**

The template should be able to load normally, with only the two "web" commands either returning blank text or printing an error to the file.

**Additional context**

(I think everything's up there)

**Comment by welpdx** [comment]
You can put those function in a try catch, to ensure that your whole script runs all the way to the end. 
```
<%*
    try {
        tR+= await tp.web.random_picture()
    } catch (err) {
        tR+= err // display error
    }
%>
```


When there is no internet, the error is `TypeError: Failed to fetch`. You can comment out `tR+= err` in the the catch block if you do not want to display this error. 

Hope that helps.

~Welp

**Comment by AB1908** [comment]
I'd recommend creating a `new Notice()` instead of adding to `tR`.

**Comment by welpdx** [comment]
> I'd recommend creating a `new Notice()` instead of adding to `tR`.

The above code meets what the guy wanted 
> either returning blank text or printing an error to the file.


But I like your idea. 

```
<%*
    try {
        tR+= await tp.web.random_picture()
    } catch (err) {
        new Notice("Error: \n" + err , 2000);
    }
%>
```

**Comment by welpdx** [comment]
@Lurux  PR #785 fixed this issue by adding try catch. 
If it has fixed your issue, please feel free to close this dicussion. Thanks!

**Comment by waltcf** [comment]
Is there any way to incorporate this behavior into Templater itself so the graceful error-catching is automatic? Having one item return some error text while the others keep working properly would be far better than everything failing due to one item's failure.

Also, how would one use the "new Notice()" approach above to insert an error text into the file itself like the first example? Or is "tR+=" the only way to do that?

**Reply by shabegom** [reply]
This should be fixed as of 1.14.1

**Reply by waltcf** [reply]
It does not appear to be, at least in my case. When I include an intentionally-bad URL in one of the web-related commands, I get the following error using Templater 1.14.3, along with the failure of the rest of the scripts in the page (including the new try/catch ones):

```
plugin:templater-obsidian:82 
Templater Error: Template parsing error, aborting. 
 Cannot read properties of undefined (reading 'periods')
```

**Reply by shabegom** [reply]
This is handled for the built in web commands. Handling a custom web request that could possibly returned undefined is out of scope.

**Reply by waltcf** [reply]
Ah, ok. Try/catch it is, then.

Is there a way to insert text about the error using the "new Notice()" command, or is "tR+=" the way to do it?

**Reply by shabegom** [reply]
You could totes use a Notice instead of tR. just replace the tR in the examples above with:

`new Notice("there was an error: "+ err)`

You might also want to `console.log(err)` to get it in the dev console. 

Also, your web request might not necessarily return an error, but the field you're trying to read could be undefined.

You should wrap the resulting code in an if/else block just in case

For example, if the object was foo.periods:

```
if (foo && foo.periods) {
//do stuff with foo.periods
} else {
new Notice("no foo for you")
}
```
That way the template can continue processing even if there is no foo.periods.
